### PR TITLE
Read config from ENV vars

### LINKED ISSUE #88: Read config from ENV vars before validation
## Challenge

We have many SAM applications deployed in many AWS accounts. As currently written, each one of our SAM templates must contain the Datadog configuration values:

```
AWSTemplateFormatVersion: "2010-09-09"
Transform:
  - AWS::Serverless-2016-10-31
  # Enable Datadog APM inside Lambda as a Lambda Layer
  - Name: DatadogServerless
    Parameters:
      stackName: !Ref "AWS::StackName"
      nodeLayerVersion: 87
      extensionLayerVersion: 40
      apiKeySecretArn: "arn:aws:secretsmanager:us-west-2:123456789012:secret:DdApiKeySecret-e1v5Yn7TvIPc-d1Qc4E"
      service: "my-service-name"
      env: !Ref Environment
```

This makes it difficult to deploy this template in an AWS account that is not `123456789012` or in a region that is not `us-west-2`. Further than this, both `apiKeySecretArn` and `env` don't actually change between SAM applications deployed in `us-west-2:123456789012` because `123456789012` is always our `dev` account. 

To make our SAM templates easily reusable across regions and accounts (without introducing Datadog specific parameters to every template), I would like to encapulate the `apiKeySecretArn` and `env` in the Macro Lambda as environment variables. The prerequisite to using this Macro (from following https://github.com/DataDog/datadog-cloudformation-macro/tree/master/serverless#installation) is to deploy the CloudFormation template `https://datadog-cloudformation-template.s3.amazonaws.com/aws/serverless-macro/latest.yml` in every account and in every region anyway, so why not make it possible to set the config at that point in time as well?

Fortunately, the code to do this is already there. [`env.ts`](https://github.com/DataDog/datadog-cloudformation-macro/blob/master/serverless/src/env.ts) is already written to use the following:
- `DD_API_KEY_SECRET_ARN` is the same as `apiKeySecretArn`
- `DD_ENV` is the same as `env`

I tried this, but unfortunately, `index.ts` validates the config supplied to the Macro before reading the environment variables for the Lambda of the macro. [Line 72](https://github.com/DataDog/datadog-cloudformation-macro/blob/a1d074d709ed60b10d646d35475134c10f7c7149/serverless/src/index.ts#L72-L80) runs and validates paramters before [Line 86](https://github.com/DataDog/datadog-cloudformation-macro/blob/a1d074d709ed60b10d646d35475134c10f7c7149/serverless/src/index.ts#L85-L86) reads the ENV vars.

## Proposal

My proposal is to modify `index.ts` to read the environment variables, which combines ENV and config params, _before_ validating them. 

However, since `setEnvConfiguration` doesn't modify `config` (it modifies [`envVariables`](https://github.com/DataDog/datadog-cloudformation-macro/blob/a1d074d709ed60b10d646d35475134c10f7c7149/serverless/src/env.ts#L216)), the validation code would need to also run against `envVariables` instead of [`config`](https://github.com/DataDog/datadog-cloudformation-macro/blob/a1d074d709ed60b10d646d35475134c10f7c7149/serverless/src/env.ts#L177). A larger refactor than just moving the validation function call, but not impossible.

## End experience

At the end of this, I could modify [`latest.yml`](https://github.com/DataDog/datadog-cloudformation-macro/blob/master/serverless/template.yml) to set the config via environment variables once per account, per region, like so:

```
Resources:
  MacroFunction:
    Type: AWS::Serverless::Function
    DependsOn: MacroFunctionZip
    Properties:
      FunctionName:
        Fn::If:
          - SetFunctionName
          - Ref: FunctionName
          - Ref: AWS::NoValue
      Description: Processes a CloudFormation template to install Datadog Lambda layers for Python and Node.js Lambda functions.
      Handler: src/index.handler
      Runtime: nodejs16.x
      CodeUri:
        Bucket: !Ref MacroFunctionZipsBucket
        Key:
          Fn::Sub:
            - "serverless-macro-${Version}.zip"
            - {
                Version:
                  !FindInMap [Constants, DatadogServerlessMacro, Version],
              }
      Environment:
        Variables:
          DD_API_KEY_SECRET_ARN: "arn:aws:secretsmanager:us-west-2:123456789012:secret:DdApiKeySecret-e1v5Yn7TvIPc-d1Qc4E"
          DD_ENV: "dev"
```

and my SAM's template would only be

```
AWSTemplateFormatVersion: "2010-09-09"
Transform:
  - AWS::Serverless-2016-10-31
  # Enable Datadog APM inside Lambda as a Lambda Layer
  - Name: DatadogServerless
    Parameters:
      stackName: !Ref "AWS::StackName"
      nodeLayerVersion: 87
      extensionLayerVersion: 40
      service: "my-service-names"
```

and because IAM policies support wildcards, the serverless function policy can be:

```
      Policies:
        - Statement:
            - Sid: GetDDAPIKey
              Effect: Allow
              Action:
                - secretsmanager:GetSecretValue
              Resource: !Sub "arn:aws:secretsmanager:${AWS::Region}:${AWS::AccountId}:secret:DdApiKeySecret-*"
```

... and work in every account and region where we choose to deploy the SAM app.

### PR DESCRIPTION
**Pull request by pas256**
### What does this PR do?

As discussed in #88, this adds the ability to specify configuration via environment variables on the Macro Lambda function.

Resolves #88

### Motivation

I wanted to simplify all of my SAM templates and not have each one specify the ARN of the Datadog secret, when it is the same in every region the Macro function is deployed.

### Testing Guidelines

The automated tests ensure the order of precedence for configuration is:
1. Default values
2. Environment variables specified on the Macro Lambda function
3. Mappings section or Parameters
... which allows the maximum flexibility.

### Additional Notes

One difference from the proposal in #88 is that I didn't have to modify any validation logic. I did this only by modifying the order in which configuration default values are read.

To be sure I didn't break anything in the tags code, I also added a test for that too.

### Types of changes

- [X] New feature

### Check all that apply

- [X] This PR's description is comprehensive
- [X] This PR impacts documentation, and it has been updated (or a ticket has been logged)
- [X] This PR's changes are covered by the automated tests

### CODE REVIEW

**Comment by pas256** [comment]
Here is my PR @sfirrin. Please let me know if you would like any modifications.

File: serverless/README.md [hunk 1]
```diff
@@ -62,6 +62,28 @@ Transform:
 
 Note: If you did not modify the provided `template.yml` file when you installed the macro, then the name of the macro defined in your account will be `DatadogServerless`. If you have modified the original template, make sure the name of the transform you add here matches the `Name` property of the `AWS::CloudFormation::Macro` resource.
 
+Note: If you want to specify some of the configuration only once, you may modify `template.yml` and add the environment variables you want to configure for that region. Think of this as a way to control additional default values. The example below sets `DD_API_KEY_SECRET_ARN` and `DD_ENV`, which the macro will treat as default values:
```
Comment (maycmlee) [inline serverless/README.md:65]:
```suggestion
Note: If you want to specify some of the configuration only once, you can modify `template.yml` and add the environment variables you want to configure for that region. This is a way to control additional default values. The example below sets `DD_API_KEY_SECRET_ARN` and `DD_ENV`, which the macro treats as default values:
```

**Review by maycmlee — COMMENTED** [review]
Thanks for opening this PR! A couple of small doc edits.

File: serverless/README.md (hunk 1 above)
Comment (pas256) [reply serverless/README.md]: Modified. Thank you!

**Review by maycmlee — APPROVED** [review]
👍 for docs

**Comment by pas256** [comment]
Amazing, thank you @sfirrin.

With this being merged now, how long before this is in the official release?

**Pull request merged**